### PR TITLE
Ckeditor integration

### PR DESCRIPTION
**Pull request by prateekmehta**
Includes Preview Test

### CODE REVIEW

**Comment by cs-shadow** [comment]
EDIT: I also checked it - Admin tests are passing locally

**Comment by cs-shadow** [comment]
Cool, the tests are all passing now. Weird why the AdminTest failed earlier.

File: js/plugins/drupalurl/plugin.js [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+              attribute = entityDOMElement.attributes.item(key);
+              attributeName = attribute.nodeName.toLowerCase();
+              if (attributeName.substring(0, 15) === 'data-cke-saved-') {
+                continue;
+              }
+              existingValues[attributeName] = existingElement.data('cke-saved-' + attributeName) || attribute.nodeValue;
+            }
+          }
+
+          //var entity_label = data.label ? data.label : existingValues['data-entity-label'];
```
Comment (cs-shadow) [inline js/plugins/drupalurl/plugin.js]: remove this stray comment

File: js/plugins/drupalurl/plugin.js [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    return null;
+  }
+
+  /**
+   * Returns whether or not the given element is a drupalurl widget.
+   *
+   * @param {CKEDITOR.editor} editor
+   * @param {CKEDITOR.htmlParser.element} element
+   */
+  function isEntityWidget (editor, element) {
```
Comment (cs-shadow) [inline js/plugins/drupalurl/plugin.js]: Change the function name (and its usages) to `isUrlWidget`

File: js/plugins/drupalurl/plugin.js [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+      delete Drupal.behaviors[behavior];
+    });
+    // Run the remaining behaviors.
+    (action == 'attach' ? Drupal.attachBehaviors : Drupal.detachBehaviors)(context, settings);
+    // Put the stashed behaviors back in.
+    $.extend(Drupal.behaviors, stashed);
+  }
+
+  /**
+   * Ajax 'url_embed_insert' command: insert the rendered entity.
```
Comment (cs-shadow) [inline js/plugins/drupalurl/plugin.js]: Should be "rendered URL"

File: js/plugins/drupalurl/plugin.js [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+   */
+  Drupal.AjaxCommands.prototype.url_embed_insert = function(ajax, response, status) {
+    var $target = ajax.element;
+    // No need to detach behaviors here, the widget is created fresh each time.
+    $target.html(response.html);
+    runEmbedBehaviors('attach', $target.get(0), response.settings || ajax.settings);
+  };
+
+  /**
+   * Stores settings specific to Entity Embed module.
```
Comment (cs-shadow) [inline js/plugins/drupalurl/plugin.js]: "URL Embed"

File: src/UrlEmbedController.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The request object.
+   * @param \Drupal\filter\FilterFormatInterface $filter_format
+   *   The filter format.
+   *
+   * @throws \Symfony\Component\HttpKernel\Exception\NotFoundHttpException
+   *   Throws an exception if 'value' parameter is not found in the request.
+   *
+   * @return \Symfony\Component\HttpFoundation\Response
+   *   The preview of the entity specified by the data attributes.
```
Comment (cs-shadow) [inline src/UrlEmbedController.php]: "of the URL"

File: js/plugins/drupalurl/plugin.js [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+   */
+  Drupal.urlEmbed = {
+    /**
+     * A list of behaviors which are to be excluded while attaching/detaching.
+     *
+     * - Drupal.behaviors.editor, to avoid CK inception.
+     * - Drupal.behaviors.contextual, to keep contextual links hidden.
+     */
+    excludedBehaviors: ['editor', 'contextual']
```
Comment (davereid) [inline js/plugins/drupalurl/plugin.js]: Not sure, but I think we can exclude these, since our output is not rendered entities, we won't have any issues with editor or contextual links in output.

File: js/plugins/drupalurl/plugin.js [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+              attributeName = attribute.nodeName.toLowerCase();
+              if (attributeName.substring(0, 15) === 'data-cke-saved-') {
+                continue;
+              }
+              existingValues[attributeName] = existingElement.data('cke-saved-' + attributeName) || attribute.nodeValue;
+            }
+          }
+
+          //var entity_label = data.label ? data.label : existingValues['data-entity-label'];
+          var url_button_id = data.id ? data.id : existingValues['url-embed-button'];
```
Comment (cs-shadow) [inline js/plugins/drupalurl/plugin.js:54]: should be `existingValues['dat-url-button']` else "edit URL" won't work.

File: js/plugins/drupalurl/plugin.js (hunk 6 above)
Comment (cs-shadow) [inline js/plugins/drupalurl/plugin.js]: Yes, I think we can remove these.

File: js/plugins/drupalurl/plugin.js [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+              attribute = entityDOMElement.attributes.item(key);
+              attributeName = attribute.nodeName.toLowerCase();
+              if (attributeName.substring(0, 15) === 'data-cke-saved-') {
+                continue;
+              }
+              existingValues[attributeName] = existingElement.data('cke-saved-' + attributeName) || attribute.nodeValue;
+            }
+          }
+
+          var url_button_id = data.id ? data.id : existingValues['url-embed-button'];
```
Comment (cs-shadow) [inline js/plugins/drupalurl/plugin.js:54]: should be existingValues['dat-url-button'] else "edit URL" won't work.

File: js/plugins/drupalurl/plugin.js [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+   * Attaches or detaches behaviors, except the ones we do not want.
+   *
+   * @param {string} action
+   *   Either 'attach' or 'detach'.
+   * @param context
+   *   The context argument for Drupal.attachBehaviors()/detachBehaviors().
+   * @param settings
+   *   The settings argument for Drupal.attachBehaviors()/detachBehaviors().
+   */
+  function runEmbedBehaviors(action, context, settings) {
```
Comment (cs-shadow) [inline js/plugins/drupalurl/plugin.js]: This function - along with only usage - can be removed.

File: src/UrlEmbedController.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  public function preview(Request $request, FilterFormatInterface $filter_format) {
+    $text = $request->get('value');
+    if ($text == '') {
+      throw new NotFoundHttpException();
+    }
+
+    $url_output = (string) check_markup($text, $filter_format->id());
+
+    $response = new AjaxResponse();
+    $response->addCommand(new UrlEmbedInsertCommand($url_output));
```
Comment (cs-shadow) [inline src/UrlEmbedController.php:47]: Since we're not using attach/detach behavior stuff, we can replace this custom command by `ReplaceCommand`. Remove `UrlEmbedInsertCommand.php` as well as `url_embed_insert` from `plugin.js`

You'll need to Preview test as well.

Comment (davereid) [inline src/UrlEmbedController.php:47]: I actually tested this earlier, and I couldn't get ReplaceCommand to actually work correctly for stuff inside the WYSIWYG. I think we might want to push forwards with the Embed module dependency so we can re-use the preview route, DomHelperTrait, and AJAX embed command.

Comment (cs-shadow) [inline src/UrlEmbedController.php:47]: I like the idea think that should do it as a follow-up otherwise this PR might get very complex to review.

Comment (cs-shadow) [inline src/UrlEmbedController.php:47]: @davereid But Embed is also using `ReplaceCommand`, right? So, how will that change anything?

**Comment by cs-shadow** [comment]
I don't know whats up with Travis, its showing random errors on builds and the errors change when I rebuild it.

File: src/UrlEmbedController.php (hunk 10 above)
Comment (davereid) [inline src/UrlEmbedController.php:47]: No, it's using a custom command. ReplaceCommand does not work for things inside iframe content.

Comment (cs-shadow) [inline src/UrlEmbedController.php:47]: Sorry, my bad. I was looking at an old version of Embed. :+1: 

Overlooked a bunch of your latest commits, so was confused.

**Comment by prateekmehta** [comment]
@davereid  Could you also please review it, so i can move forward with filed formatter.

**Comment by davereid** [comment]
@prateekmehta The field formatter can be done completely independently of any of this work, please do not let it hold you up.

**Comment by davereid** [comment]
It would be nice to get a rebase on all of these commits.

**Comment by cs-shadow** [comment]
@davereid Seems like they already are rebased.

**Comment by davereid** [comment]
Grr, Github not updating. Thanks, I see that now.
